src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/exceptions.kt

```diff
@@ -0,0 +1,24 @@
+/**
+    Copyright 2018 BarD Software s.r.o
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Author: Mikhail Shavkunov (@shavkunov)
+ */
+package com.bardsoftware.backend.fleet.rmarkdown
```

Comment:
I see you like exceptions. 
Please follow file naming conventions though. File name must be capitalized.

Besides, such variety of exceptions is rarely needed. You basically do not care if directory does not exist or is not writable, all you want to do when your subscriber can't write to the directory is to fail. Informative message is more important than exception class. You can just throw generic IOException. If its message contains the reason and _the name_ of the non-existing/non-writable directory, it will be enough for the operator who reads the logs.

The same with RootFileNotFoundException. If subscriber decides that it can't proceed with the compilation, it should somehow notify those who may be concerned, but most likely when you debug you do not care about exception class and more interested in the details of failure, e.g. what is the (full) path of the missing root file. Is there any error in the unpacking code probably, and it tries to find a file which is not supposed to exist?

Reply:
Indeed, all the exceptions can be thrown as IO exception with cause comment. It's just a habit from java course :)
Done